libengine/dpengine/pixels.c

```diff
@@ -431,6 +433,137 @@ void DP_pixels15_to_8(DP_Pixel8 *dst, const DP_Pixel15 *src, int count)
     }
 }
 
+#ifdef DP_CPU_X64
+DP_TARGET_BEGIN("sse4.2")
+static void DP_pixels15_to_8_sse42(DP_Pixel8 *dst, const DP_Pixel15 *src,
+                                   int count)
+{
+    assert(count % 4 == 0);
+
+    for (int i = 0; i < count; i += 4) {
+        __m128i source1 = _mm_load_si128((__m128i *)&src[i]);
+        __m128i source2 = _mm_load_si128((__m128i *)&src[i + 2]);
+
+        __m128i p1 = _mm_cvtepu16_epi32(source1);
+        __m128i p2 = _mm_cvtepu16_epi32(_mm_srli_si128(source1, 8));
+        __m128i p3 = _mm_cvtepu16_epi32(source2);
+        __m128i p4 = _mm_cvtepu16_epi32(_mm_srli_si128(source2, 8));
+
+        p1 = _mm_srli_epi32(_mm_mullo_epi32(p1, _mm_set1_epi32(255)), 15);
+        p2 = _mm_srli_epi32(_mm_mullo_epi32(p2, _mm_set1_epi32(255)), 15);
+        p3 = _mm_srli_epi32(_mm_mullo_epi32(p3, _mm_set1_epi32(255)), 15);
+        p4 = _mm_srli_epi32(_mm_mullo_epi32(p4, _mm_set1_epi32(255)), 15);
+
+        p2 = _mm_slli_si128(p2, 1);
+        p3 = _mm_slli_si128(p3, 2);
+        p4 = _mm_slli_si128(p4, 3);
+
+        __m128i combined =
+            _mm_or_si128(p1, _mm_or_si128(p2, _mm_or_si128(p3, p4)));
+
+        __m128i out = _mm_shuffle_epi8(
+            combined, _mm_setr_epi8(0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14, 3,
+                                    7, 11, 15));
+
+        _mm_store_si128((__m128i *)&dst[i], out);
+    }
+}
+DP_TARGET_END
+
+DP_TARGET_BEGIN("avx2")
+static void DP_pixels15_to_8_avx2(DP_Pixel8 *dst, const DP_Pixel15 *src,
+                                  int count)
+{
+    assert(count % 8 == 0);
+
+    for (int i = 0; i < count; i += 8) {
+        __m256i source1 = _mm256_loadu_si256((__m256i *)&src[i]);
+        __m256i source2 = _mm256_loadu_si256((__m256i *)&src[i + 4]);
+
+        __m256i p1 = _mm256_and_si256(source1, _mm256_set1_epi32(0xFFFF));
+        __m256i p2 = _mm256_srli_epi32(
+            _mm256_and_si256(source1, _mm256_set1_epi32(0xFFFF0000)), 16);
+        __m256i p3 = _mm256_and_si256(source2, _mm256_set1_epi32(0xFFFF));
+        __m256i p4 = _mm256_srli_epi32(
+            _mm256_and_si256(source2, _mm256_set1_epi32(0xFFFF0000)), 16);
```

Comment:
Hex is usually written with lowercase letters in Drawdance.

The two large foo values needs a cast to avoid a warning about unsigned to signed conversion: `(int)0xfff0000`